ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,467 @@
+# ENSIP-15: ENS Name Normalization Standard
+
+| **Author**  | Andrew Raffensperger \<raffy@me.com> |
+| ----------- | ------------------------------------ |
+| **Status**  | Draft                                |
+| **Created** | 2023-04-03                           |
+
+## Abstract
+
+This ENSIP standardizes Ethereum Name Service (ENS) name normalization process outlined in [ENSIP-1 § Name Syntax](./ensip-1-ens.md#name-syntax).
+
+## Motivation
+
+* Since [ENSIP-1](./ensip-1-ens.md) (originally [EIP-137](https://eips.ethereum.org/EIPS/eip-137)) was finalized in 2016, Unicode has [evolved](https://unicode.org/history/publicationdates.html) from version 8.0.0 to 15.0.0 and incorporated many new characters, including complex emoji sequences. 
+* ENSIP-1 does not state the version of Unicode.
+* ENSIP-1 implies but does not state an explicit flavor of IDNA processing. 
+* [UTS-46](https://unicode.org/reports/tr46/) is insufficient to normalize emoji sequences. Correct emoji processing is only possible with [UTS-51](https://www.unicode.org/reports/tr51/).
+* Validation tests are needed to ensure implementation compliance.
+* The success of ENS has encouraged spoofing via the following techniques:
+	1. Insertion of zero-width characters.
+	1. Using names which normalize differently between algorithms. 
+	1. Using names which appear differently between applications and devices.
+	1. Substitution of confusable (look-alike) characters.
+	1. Mixing incompatible scripts.
+
+## Specification
+
+* Unicode version `15.0.0`
+	* Normalization is a living specification and should use the latest stable version of Unicode.
+* [`spec.json`](./ensip-15/spec.json) contains all [necessary data](#description-of-specjson) for normalization.
+* [`nf.json`](./ensip-15/nf.json) contains all necessary data for [Unicode Normalization Forms](https://unicode.org/reports/tr15/) NFC and NFD.

```

Comment:
How exactly is `nf.json` supposed to be used to aid in the implementation? How does an implementer reference the `ranks` key and how is one supposed to interpret the lists in there?

I'm curious if this can actually be used as a reference with static value mapping (as `spec.json`  tends to help with) or if during implementation `NFC` and / or `NFD` actually need to be applied to the text tokens.

An explainer of this file would be quite helpful.

Reply:
You are correct, I didn't provide an explanation.

My thinking was that NF version issue is really only important on the web, where there are almost no space-efficient client-side Unicode libraries, and my library and the official ENS library cover that use-case.  I was assuming any non-web user would just import a library with the matching Unicode version.

I will include a section in my next update.

---

For reference, `ranks` is just a dense packing of Unicode [combining classes.](https://www.compart.com/en/unicode/combining)

Because NF orders combining marks, a static replacement table is insufficient.

Reply:
Makes sense. Thanks for the prompt reply 👌🏼